persistence/sql/persister_recovery.go

```diff
@@ -129,66 +131,127 @@ func (p *Persister) DeleteExpiredRecoveryFlows(ctx context.Context, expiresAt ti
 	return nil
 }
 
-func (p *Persister) CreateRecoveryCode(ctx context.Context, recoveryCode *code.RecoveryCode) error {
+func (p *Persister) CreateRecoveryCode(ctx context.Context, dto *code.RecoveryCodeDTO) (*code.RecoveryCode, error) {
 	ctx, span := p.r.Tracer(ctx).Tracer().Start(ctx, "persistence.sql.CreateRecoveryCode")
 	defer span.End()
 
-	code := recoveryCode.Code
-	recoveryCode.Code = p.hmacValue(ctx, code)
-	recoveryCode.NID = p.NetworkID(ctx)
+	now := time.Now()
+
+	recoveryCode := &code.RecoveryCode{
+		ID:         x.NewUUID(),
+		Code:       p.hmacValue(ctx, dto.Code),
+		ExpiresAt:  now.UTC().Add(dto.ExpiresIn),
+		IssuedAt:   now,
+		CodeType:   dto.CodeType,
+		FlowID:     dto.FlowID,
+		NID:        p.NetworkID(ctx),
+		IdentityID: dto.IdentityID,
+	}
+
+	if dto.RecoveryAddress != nil {
+		recoveryCode.RecoveryAddress = dto.RecoveryAddress
+		recoveryCode.RecoveryAddressID = uuid.NullUUID{
+			UUID:  dto.RecoveryAddress.ID,
+			Valid: true,
+		}
+	}
 
 	// This should not create the request eagerly because otherwise we might accidentally create an address that isn't
 	// supposed to be in the database.
 	if err := p.GetConnection(ctx).Create(recoveryCode); err != nil {
-		return err
+		return nil, err
 	}
 
-	recoveryCode.Code = code
-	return nil
+	return recoveryCode, nil
 }
 
 func (p *Persister) UseRecoveryCode(ctx context.Context, fID uuid.UUID, codeVal string) (*code.RecoveryCode, error) {
 	ctx, span := p.r.Tracer(ctx).Tracer().Start(ctx, "persistence.sql.UseRecoveryCode")
 	defer span.End()
 
-	var recoveryCode code.RecoveryCode
+	var recoveryCode *code.RecoveryCode
 
 	nid := p.NetworkID(ctx)
+
+	flowTableName := new(recovery.Flow).TableName(ctx)
+
 	if err := sqlcon.HandleError(p.Transaction(ctx, func(ctx context.Context, tx *pop.Connection) (err error) {
+
+		/* #nosec G201 TableName is static */
+		if err = sqlcon.HandleError(tx.RawQuery(fmt.Sprintf("UPDATE %s SET submit_count = submit_count + 1 WHERE id = ? AND nid = ?", flowTableName), fID, nid).Exec()); err != nil {
+			return
+		}
+
+		var submitCount int
+		// Because MySQL does not support "RETURNING" clauses, but we need the updated `submit_count` later on.
+		/* #nosec G201 TableName is static */
+		if err := sqlcon.HandleError(tx.RawQuery(fmt.Sprintf("SELECT submit_count FROM %s WHERE id = ? AND nid = ?", flowTableName), fID, nid).First(&submitCount)); err != nil {
+			if errors.Is(err, sqlcon.ErrNoRows) {
+				// Return no error, as that would roll back the transaction
+				return nil
+			}
+
+			return err
+		}
+
+		if submitCount > 5 {
+			return code.ErrCodeSubmittedTooOften
+		}
+
+		var recoveryCodes []code.RecoveryCode
+		if err = sqlcon.HandleError(tx.Where("nid = ? AND selfservice_recovery_flow_id = ?", nid, fID).All(&recoveryCodes)); err != nil {
```

Comment:
Do we expect to have more then one recovery code per flow?

Reply:
Yes, "resend" is one case where this can happen.

Reply:
Would that invalidate the older code?

Reply:
No, it does not, but I also thought of that. Maybe we should do that?

UX: 
- Gmail sometimes groups these emails together, which makes it hard to distinguish between the first and second email, so it would be nice if both codes work

Security:
- Having multiple codes does open up the possibility to brute force these codes. Especially, since we don't have a limit on codes. That might also be an option: `if "count(codes) > 5" then "retry the flow"` e.g. if a user clicked "resend" more than 5 times, we create a new flow. 

Overall, I think security is more important here, and the easy fix is to just delete all codes from the database if a new code is requested. WDYT?

Reply:
> Overall, I think security is more important here, and the easy fix is to just delete all codes from the database if a new code is requested. WDYT?

Let's do that!

Reply:
Done.